### PR TITLE
Address Query Parsing Performance

### LINKED ISSUE #4394: Complex Query Takes Forever - but I think it's query generation
### Describe your issue

I've a very complex query, wich Takes forever to run (over 1 Minute. without any data in the database).
But for me it looks like the query generation is taking so long.
It needs #4339 merged to work in current linq2db

### Steps to reproduce

I've created an sample - https://github.com/linq2db/linq2db/pull/4395

### Environment details

`Linq To DB` version: master

Database (with version): Sqlite and MSSql

### PR DESCRIPTION
**Pull request by jogibear9988**
Fixes #4394

### CODE REVIEW

**Comment by MaceWindu** [comment]
image

**Comment by MaceWindu** [comment]
and in release
image

**Comment by jogibear9988** [comment]
I merged #4339 in this pull request, cause without the query would not run at all.

**Comment by MaceWindu** [comment]
I doubt it makes sense to do anything here before new parser lends. Profiler shows about half of time spent in this visitor https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/SqlQuery/SelectQueryOptimizer.cs#L455-L468

@sdanyliv ?

**Comment by MaceWindu** [comment]
image

**Comment by viceroypenguin** [comment]
> I doubt it makes sense to do anything here before new parser lends. Profiler shows about half of time spent in this method https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/SqlQuery/SelectQueryOptimizer.cs#L453-L482
> 
> @sdanyliv ?

To be specific, it's in the lambda from L455-L468

**Comment by MaceWindu** [comment]
new parser shows similar numbers currently, but as I said - it's better leave it for later as slow areas will be heavy refactored there

**Comment by igor-tkachev** [comment]
> new parser shows similar numbers currently, but as I said - it's better leave it for later as slow areas will be heavy refactored there

This is a bug that has to be fixed ASAP. We don't know how long the new parser will still be in development and how long it will take us to fix bugs in it. Will this bug be present in the library all this time?

**Comment by igor-tkachev** [comment]
I like this test. Profiler shows a lot of points to optimize.

**Comment by sdanyliv** [comment]
Profiler adds additional delays. Without profiler in Release, whole query is executed under 550-600 milliseconds. Second run with cache hit is 8 msec. There are a lot of ways how to improve SelectQuery optimization  and I will make research for new LINQ Translator but after all tests are passed. Not too much left.

Note that SelectQuery optimization is fully REWRITTEN, making cardinal changes in current version will be discarded .

**Review by sdanyliv — CHANGES_REQUESTED** [review]

**Comment by jogibear9988** [comment]
@sdanyliv 

Your coments were already resolved in pull req for issue #4337, wich I needed, without this didn't run at all

File: Tests/Linq/UserTests/Issue4337Tests.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@
 
 using NUnit.Framework;
 
+using static Tests.UserTests.Issue4394Tests;
```
Comment (sdanyliv) [inline Tests/Linq/UserTests/Issue4337Tests.cs:12]: Remove that

**Comment by sdanyliv** [comment]
Tested on V6
In Release second query executes approx 8 seconds without Stack Overflow. In Debug it crashes with SO, Probably too much debug information

Good test for optimizing translator, bad for production. 

Your goal to kill `linq2db` or make job done?

**Pull request merged**